experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala

```diff
@@ -0,0 +1,344 @@
+
+object PipelineConfig {
+  implicit def getConfig(pipelineConfig: PipelineConfig): Config = pipelineConfig.config
+}
+
+object DatumHandler {
+  val DEFAULT_INTERVAL = 2
+  val CPU = "CPU"
+  val CPU_INTERVAL = "pipeline.cpu.interval"
+  val MEM = "MEM"
+  val MEM_INTERVAL = "pipeline.memory.interval"
+
+  implicit def handle(data: String): Datum = {
+    read[Datum](data)
+  }
+  implicit def handle(datum: Datum): String = {
+    write[Datum](datum)
+  }
+}
+
+trait Average {
+  this: Task =>
+  var averageMem: Double = 0
+  var totalCount: Long = 0
+  var totalSum: Double = 0
+  var timeStamp: TimeStamp = 0
+
+  def timeInterval: Int
+
+  def average(datum: Datum)(implicit timeStamp: TimeStamp): Option[Message] = {
+    totalCount += 1
+    totalSum += datum.value
+    averageMem = totalSum/totalCount
+    interval(datum)
+  }
+
+  def elapsedInSec(implicit ts: Long): Long = ts - timeStamp
+
+  def interval(datum: Datum)(implicit ts: TimeStamp): Option[Message] = {
+    timeStamp match {
+      case 0 =>
+        timeStamp = ts
+        None
+      case _ =>
+        elapsedInSec match {
+          case delta if delta > timeInterval =>
+            timeStamp = ts
+            Some(Message(write[Datum](Datum(datum.dimension,datum.metric,averageMem)), timeStamp))
+          case _ =>
+            None
+        }
+    }
+  }
+}
+
+class CpuProcessor(taskContext: TaskContext, conf: UserConfig)
+  extends Task(taskContext, conf) with Average {
+  import taskContext.output
+
+  val pipelineConfig = taskConfig.getValue[PipelineConfig](PIPELINE)
+  override def timeInterval = pipelineConfig.map(config => {
+    config.getInt(CPU_INTERVAL)
+  }).getOrElse(DEFAULT_INTERVAL)
+
+  override def onStart(newStartTime: StartTime): Unit = {
+    LOG.info("starting timeInterval=$timeInterval")
+  }
+
+  override def onNext(msg: Message): Unit = {
+    Try({
+      implicit val timeStamp = msg.timestamp
+      val jsonData = msg.msg.asInstanceOf[String]
+      val envelope = read[Envelope](jsonData)
+      val body = read[Body](envelope.body)
+      val metrics = body.metrics
+      metrics.foreach(datum => {
+        datum.dimension match {
+          case CPU =>
+            average(datum).foreach(output)
+          case _ =>
+        }
+      })
+    }).failed.foreach(LOG.error("bad message", _))
+  }
+}
+
+class MemoryProcessor(taskContext: TaskContext, conf: UserConfig)
+  extends Task(taskContext, conf) with Average {
+  import taskContext.output
+
+  val pipelineConfig = taskConfig.getValue[PipelineConfig](PIPELINE)
+  override def timeInterval = pipelineConfig.map(config => {
+    config.getInt(MEM_INTERVAL)
+  }).getOrElse(DEFAULT_INTERVAL)
+
+  override def onStart(newStartTime: StartTime): Unit = {
+    LOG.info("starting timeInterval=$timeInterval")
+  }
+
+  override def onNext(msg: Message): Unit = {
+    Try({
+      implicit val timeStamp = msg.timestamp
+      val jsonData = msg.msg.asInstanceOf[String]
```

Comment:
Tried adding this to pipeline.conf

gearpump {
  serializers {
    "scala.Array" = ""
    "org.apache.gearpump.experiments.pipeline.Messages$Envelope" = ""
    "org.apache.gearpump.experiments.pipeline.Messages$Body" = ""
    "org.apache.gearpump.experiments.pipeline.Messages$Datum" = ""
  }
}

Getting this error:

[ERROR] [04/01/2015 16:52:45.931] [KafkaProducer@app1exec0TaskId(0,0)] caught error
java.lang.ClassCastException: java.lang.String cannot be cast to org.apache.gearpump.experiments.pipeline.Messages$Envelope
        at org.apache.gearpump.experiments.pipeline.KafkaProducer$$anonfun$onNext$5$$anonfun$apply$mcV$sp$3.apply(Processors.scala:241)
        at org.apache.gearpump.experiments.pipeline.KafkaProducer$$anonfun$onNext$5.apply$mcV$sp(Processors.scala:240)
        at org.apache.gearpump.experiments.pipeline.KafkaProducer$$anonfun$onNext$5.apply(Processors.scala:240)
        at org.apache.gearpump.experiments.pipeline.KafkaProducer$$anonfun$onNext$5.apply(Processors.scala:240)
        at scala.util.Try$.apply(Try.scala:191)
        at org.apache.gearpump.experiments.pipeline.KafkaProducer.onNext(Processors.scala:239)
        at org.apache.gearpump.streaming.task.TaskWrapper.onNext(TaskWrapper.scala:84)
        at org.apache.gearpump.streaming.task.TaskActor.onNext(TaskActor.scala:82)
        at org.apache.gearpump.streaming.task.TaskActor.org$apache$gearpump$streaming$task$TaskActor$$doHandleMessage(TaskActor.scala:212)
        at org.apache.gearpump.streaming.task.TaskActor$$anonfun$stashAndHandleMessages$1.applyOrElse(TaskActor.scala:271)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
        at org.apache.gearpump.streaming.task.TaskActor.aroundReceive(TaskActor.scala:36)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
        at akka.actor.ActorCell.invoke(ActorCell.scala:487)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
        at akka.dispatch.Mailbox.run(Mailbox.scala:221)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

Also tried

gearpump {
  serializers {
    "scala.Array" = ""
    "org.apache.gearpump.experiments.pipeline.Messages.Envelope" = ""
    "org.apache.gearpump.experiments.pipeline.Messages.Body" = ""
    "org.apache.gearpump.experiments.pipeline.Messages.Datum" = ""
  }
}